File: java/abc204/C.java
import java.util.*;

public class C {
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int n = sc.nextInt();
    int m = sc.nextInt();
    int[][] ab = new int[m][2];
    for (int i = 0; i < m; i++) {
      Arrays.setAll(ab[i], j -> sc.nextInt());
    }

    HashMap<Integer, ArrayList<Integer>> hm = new HashMap<>();
    for (int i = 0; i < m; i++) {
      var entry = hm.get(ab[i][0]);
      if (entry == null) {
        ArrayList<Integer> al = new ArrayList<>();
        al.add(ab[i][1]);
        hm.put(ab[i][0], al);
      } else {
        entry.add(ab[i][1]);
        hm.put(ab[i][0], entry);
      }
    }

    int cnt = 0;
    for (int i = 1; i < n + 1; i++) {
      HashSet<Integer> stamp = new HashSet<>();
      stamp.add(i);
      findNext(i, stamp, hm);
      cnt += stamp.size();
    }

    System.out.println(cnt);
  }

  public static void findNext(int current, HashSet<Integer> stamp, HashMap<Integer, ArrayList<Integer>> hm) {
    var nexts = hm.get(current);
    if (nexts != null) {
      for (Integer p : nexts) {
        if (!stamp.contains(p)) {
          stamp.add(p);
          findNext(p, stamp, hm);
        }
      }
    }
  }
}

File: rust/abc204/src/c.rs
use proconio::input;
use std::collections::{HashMap, HashSet};
use std::io::{stdout, BufWriter, Write};

pub fn main() {
  let out = stdout();
  let mut out = BufWriter::new(out.lock());

  input! {n: u32, m: u32, ab: [(u32, u32); m]}

  let mut hm: HashMap<u32, Vec<u32>> = HashMap::new();
  for road in ab {
    let hm_entry = hm.entry(road.0).or_insert(Vec::new());
    hm_entry.push(road.1)
  }

  let mut cnt = 0;
  for start in 1..n + 1 {
    let mut stamp: HashSet<u32> = HashSet::new();
    stamp.insert(start);
    find_nexts(&start, &mut stamp, &hm);
    cnt += stamp.len();
  }
  writeln!(out, "{}", cnt).unwrap();
}

fn find_nexts(current: &u32, stamp: &mut HashSet<u32>, hm: &HashMap<u32, Vec<u32>>) {
  if let Some(nexts) = hm.get(&current) {
    for node in nexts {
      if !stamp.contains(node) {
        stamp.insert(*node);
        find_nexts(node, stamp, hm);
      }
    }
  }
}
